### PR TITLE
Separate FAB migration from Core Airflow migration

### PR DESCRIPTION
**Pull request by ephraimbuddy**
This PR separates FAB migration from Airflow Core migration and provides a way for apps to integrate into Airflow and run their migrations.

### CODE REVIEW

**Review by potiuk — COMMENTED** [review]
NICE

**Review by vincbeck — APPROVED** [review]
🤯 Fantastic! I wish I could have done it while working on AIP-56! Thank you!

**Comment by ephraimbuddy** [comment]
Not sure why this test is failing but doesn't fail locally: https://github.com/apache/airflow/actions/runs/10406470435/job/28819881902?pr=41437#step:7:5512

**Comment by ephraimbuddy** [comment]
> Not sure why this test is failing but doesn't fail locally: https://github.com/apache/airflow/actions/runs/10406470435/job/28819881902?pr=41437#step:7:5512

Mostly tests related to logging: 
https://github.com/apache/airflow/actions/runs/10406470435/job/28819876941?pr=41437#step:7:5339

**Comment by potiuk** [comment]
Likely side-effect of other tests that have been somewhat masked or avoided before the change - where some setup/teardown removed the side-effect.

You can reproduce the set of tests run with `breeze testing db-tests --test-type CLI` locally (for example - for CLI tests) - that should run the tests in the same sequence as they are run in CI in each of the parallel runs and then they shoudl reproducibly fail as well. Also it could be caused by new version of dependencies - (look at `generate constraints` output of your build) but it's rather unlikely - https://github.com/apache/airflow/actions/runs/10420774982/ `canary` build just got green and updated constraints without any test failures, so it's rather unlikely (however you can always rebase and see if it will fail in the same way).

Generally those kind of side-effects are best investigated by a bit guessing and bi-secting - and trying to run a smaller-and-smaller subset of tests until you find the one that is the culprit. At least that's what I did in the past.

You should start by looking  at the pytest command that was run in the original test type - unfold the `red` failing test type and you will see:

```
  Starting the tests with those pytest arguments: tests/cli --verbosity=0 --strict-markers --durations=100 --maxfail=50 --color=yes --junitxml=/files/test_result-cli-postgres.xml --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 --disable-warnings -rfEX --run-db-tests-only --ignore=tests/system --ignore=tests/integration --warning-output-path=/files/warnings-cli-postgres.txt --ignore=helm_tests --with-db-init --no-cov
  
  ============================= test session starts ==============================
  platform linux -- Python 3.8.19, pytest-8.3.2, pluggy-1.5.0
  rootdir: /opt/airflow
  configfile: pyproject.toml
  plugins: icdiff-0.9, timeouts-1.2.1, instafail-0.5.0, custom-exit-code-0.3.0, rerunfailures-14.0, asyncio-0.23.8, time-machine-2.15.0, anyio-4.4.0, requests-mock-1.12.1, cov-5.0.0, mock-3.14.0, xdist-3.6.1
  asyncio: mode=strict
  setup timeout: 60.0s, execution timeout: 60.0s, teardown timeout: 60.0s
  collected 406 items
  
  tests/cli/commands/test_celery_command.py ..........                     [  2%]
  tests/cli/commands/test_cheat_sheet_command.py s                         [  2%]
  tests/cli/commands/test_config_command.py ssssssssssssssssss             [  7%]
  tests/cli/commands/test_connection_command.py .......................... [ 13%]
  .....................                                                    [ 18%]
  tests/cli/commands/test_dag_command.py ................................. [ 26%]
  ....................                                                     [ 31%]
  tests/cli/commands/test_dag_processor_command.py .                       [ 32%]
  tests/cli/commands/test_db_command.py .................................. [ 40%]
  .....................................                                    [ 49%]
  tests/cli/commands/test_info_command.py sssssssss..s                     [ 52%]
  tests/cli/commands/test_internal_api_command.py ssss...                  [ 54%]
  tests/cli/commands/test_jobs_command.py ......                           [ 55%]
  tests/cli/commands/test_kerberos_command.py ....                         [ 56%]
  tests/cli/commands/test_kubernetes_command.py ..........                 [ 59%]
  tests/cli/commands/test_legacy_commands.py sss                           [ 59%]
  tests/cli/commands/test_plugins_command.py ...                           [ 60%]
  tests/cli/commands/test_pool_command.py ...........                      [ 63%]
  tests/cli/commands/test_rotate_fernet_key_command.py ..                  [ 63%]
  tests/cli/commands/test_scheduler_command.py ...................         [ 68%]
  tests/cli/commands/test_standalone_command.py ssssssssssssss             [ 71%]
  tests/cli/commands/test_task_command.py ................................ [ 79%]
  .F............                                                           [ 83%]
  tests/cli/commands/test_triggerer_command.py ..                          [ 83%]
  tests/cli/commands/test_variable_command.py ...........                  [ 86%]
  tests/cli/commands/test_version_command.py s                             [ 86%]
  tests/cli/commands/test_webserver_command.py sssssssssss....             [ 90%]
  tests/cli/test_cli_parser.py ..................................s....     [100%]
```

If your tests succeeds when run separately, but fails when run as `tests/cli` - then side-effect is almost certain root cause. And you can attempt guess which one is producing the side-effect and run only that test and the one that's failing to confirm your guess. Or attempt to bisect it:

In this case you might convert the single:

* `pytest --run-db-tests-only tests/cli`  (that should fail locally for you as well)

into (looking at the output): 

* `pytest --run-db-tests-only  tests/cli/commands/test_celery_command.py tests/cli/commands/test_cheat_sheet_command.py ... tests/cli/commands/test_task_command.py` 

Then you can remove half of the modules from the list and run it again (then you will see whether side-effect comes from the removed half or the remaining half). And continue that path - even down to a single test that causes the side effect. Then usually fixing it is trivial by adding missing setup/teardown or changing the test so that it patches and restores any state.

It's slow and tedious, yes, but this is the way I've been successfully using in the past to trace root causes of similar issues, and have no other idea how to do it differently faster.

**Comment by ephraimbuddy** [comment]
> Likely side-effect of other tests that have been somewhat masked or avoided before the change - where some setup/teardown removed the side-effect.
> 
> 
> 
> You can reproduce the set of tests run with `breeze testing db-tests --test-type CLI` locally (for example - for CLI tests) - that should run the tests in the same sequence as they are run in CI in each of the parallel runs and then they shoudl reproducibly fail as well. Also it could be caused by new version of dependencies - (look at `generate constraints` output of your build) but it's rather unlikely - https://github.com/apache/airflow/actions/runs/10420774982/ `canary` build just got green and updated constraints without any test failures, so it's rather unlikely (however you can always rebase and see if it will fail in the same way).
> 
> 
> 
> Generally those kind of side-effects are best investigated by a bit guessing and bi-secting - and trying to run a smaller-and-smaller subset of tests until you find the one that is the culprit. At least that's what I did in the past.
> 
> 
> 
> You should start by looking  at the pytest command that was run in the original test type - unfold the `red` failing test type and you will see:
> 
> 
> 
> ```
> 
>   Starting the tests with those pytest arguments: tests/cli --verbosity=0 --strict-markers --durations=100 --maxfail=50 --color=yes --junitxml=/files/test_result-cli-postgres.xml --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 --disable-warnings -rfEX --run-db-tests-only --ignore=tests/system --ignore=tests/integration --warning-output-path=/files/warnings-cli-postgres.txt --ignore=helm_tests --with-db-init --no-cov
> 
>   
> 
>   ============================= test session starts ==============================
> 
>   platform linux -- Python 3.8.19, pytest-8.3.2, pluggy-1.5.0
> 
>   rootdir: /opt/airflow
> 
>   configfile: pyproject.toml
> 
>   plugins: icdiff-0.9, timeouts-1.2.1, instafail-0.5.0, custom-exit-code-0.3.0, rerunfailures-14.0, asyncio-0.23.8, time-machine-2.15.0, anyio-4.4.0, requests-mock-1.12.1, cov-5.0.0, mock-3.14.0, xdist-3.6.1
> 
>   asyncio: mode=strict
> 
>   setup timeout: 60.0s, execution timeout: 60.0s, teardown timeout: 60.0s
> 
>   collected 406 items
> 
>   
> 
>   tests/cli/commands/test_celery_command.py ..........                     [  2%]
> 
>   tests/cli/commands/test_cheat_sheet_command.py s                         [  2%]
> 
>   tests/cli/commands/test_config_command.py ssssssssssssssssss             [  7%]
> 
>   tests/cli/commands/test_connection_command.py .......................... [ 13%]
> 
>   .....................                                                    [ 18%]
> 
>   tests/cli/commands/test_dag_command.py ................................. [ 26%]
> 
>   ....................                                                     [ 31%]
> 
>   tests/cli/commands/test_dag_processor_command.py .                       [ 32%]
> 
>   tests/cli/commands/test_db_command.py .................................. [ 40%]
> 
>   .....................................                                    [ 49%]
> 
>   tests/cli/commands/test_info_command.py sssssssss..s                     [ 52%]
> 
>   tests/cli/commands/test_internal_api_command.py ssss...                  [ 54%]
> 
>   tests/cli/commands/test_jobs_command.py ......                           [ 55%]
> 
>   tests/cli/commands/test_kerberos_command.py ....                         [ 56%]
> 
>   tests/cli/commands/test_kubernetes_command.py ..........                 [ 59%]
> 
>   tests/cli/commands/test_legacy_commands.py sss                           [ 59%]
> 
>   tests/cli/commands/test_plugins_command.py ...                           [ 60%]
> 
>   tests/cli/commands/test_pool_command.py ...........                      [ 63%]
> 
>   tests/cli/commands/test_rotate_fernet_key_command.py ..                  [ 63%]
> 
>   tests/cli/commands/test_scheduler_command.py ...................         [ 68%]
> 
>   tests/cli/commands/test_standalone_command.py ssssssssssssss             [ 71%]
> 
>   tests/cli/commands/test_task_command.py ................................ [ 79%]
> 
>   .F............                                                           [ 83%]
> 
>   tests/cli/commands/test_triggerer_command.py ..                          [ 83%]
> 
>   tests/cli/commands/test_variable_command.py ...........                  [ 86%]
> 
>   tests/cli/commands/test_version_command.py s                             [ 86%]
> 
>   tests/cli/commands/test_webserver_command.py sssssssssss....             [ 90%]
> 
>   tests/cli/test_cli_parser.py ..................................s....     [100%]
> 
> ```
> 
> 
> 
> If your tests succeeds when run separately, but fails when run as `tests/cli` - then side-effect is almost certain root cause. And you can attempt guess which one is producing the side-effect and run only that test and the one that's failing to confirm your guess. Or attempt to bisect it:
> 
> 
> 
> In this case you might convert the single:
> 
> 
> 
> * `pytest --run-db-tests-only tests/cli`  (that should fail locally for you as well)
> 
> 
> 
> into (looking at the output): 
> 
> 
> 
> * `pytest --run-db-tests-only  tests/cli/commands/test_celery_command.py tests/cli/commands/test_cheat_sheet_command.py ... tests/cli/commands/test_task_command.py` 
> 
> 
> 
> Then you can remove half of the modules from the list and run it again (then you will see whether side-effect comes from the removed half or the remaining half). And continue that path - even down to a single test that causes the side effect. Then usually fixing it is trivial by adding missing setup/teardown or changing the test so that it patches and restores any state.
> 
> 
> 
> It's slow and tedious, yes, but this is the way I've been successfully using in the past to trace root causes of similar issues, and have no other idea how to do it differently faster.

Thanks @potiuk for always helping. I'll look into these and fix them 🙏

File: airflow/providers/fab/__init__.py [hunk 1]
```diff
@@ -32,8 +32,8 @@
 __version__ = "1.3.0"
 
 if packaging.version.parse(packaging.version.parse(airflow_version).base_version) < packaging.version.parse(
-    "2.9.0"
+    "3.0.0"
```
Comment (potiuk) [inline airflow/providers/fab/__init__.py]: Hm. That would mean that we won't have FAB provider released for Airflow 2 and I think it's not good, because for example new FAB might get released with critical security fixes and we won't be able to upgrade it

I think we should find a solution with compatibility layer for Airlfow 2 - either via some conditional imports in FAB or implementing some `airflow.providers.compat` schim.

Comment (ephraimbuddy) [reply airflow/providers/fab/__init__.py]: Sure. I was trying to see if the compact test failures would be fixed. I will remove it and implement proper compatibility shim

**Comment by eladkal** [comment]
wait.. what does this actually mean for users who is bumping providers only?
We have doc that explain upgrade procedure https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html#upgrading-airflow-to-a-newer-version

If now, migrations can also run from bumping provider only that changes the upgrade procedure users should take.

**Comment by ephraimbuddy** [comment]
> wait.. what does this actually mean for users who is bumping providers only? We have doc that explain upgrade procedure https://airflow.apache.org/docs/apache-airflow/stable/installation/upgrading.html#upgrading-airflow-to-a-newer-version
> 
> If now, migrations can also run from bumping provider only that changes the upgrade procedure users should take.

I will describe the upgrade procedure in my next PR when I add the upgrade command for FAB provider. It should be smooth

**Comment by eladkal** [comment]
> I will describe the upgrade procedure in my next PR when I add the upgrade command for FAB provider. It should be smooth

I am worried here.
We may dismiss the impact here too lightly.
Did I miss mailing thread on this topic?

Here we are introducing something really new - bumping provider version which also runs DB migration.. that is not small change to how Airflow operatre.

**Comment by ephraimbuddy** [comment]
> > I will describe the upgrade procedure in my next PR when I add the upgrade command for FAB provider. It should be smooth
> 
> I am worried here. We may dismiss the impact here too lightly. Did I miss mailing thread on this topic?
> 
> Here we are introducing something really new - bumping provider version which also runs DB migration.. that is not small change to how Airflow operatre.

These changes are part of AIP-79, do you think we still need another mailing thread on it?

**Comment by eladkal** [comment]
> These changes are part of AIP-79, do you think we still need another mailing thread on it?

To clarify, I am not saying no to the what (the merit of AIP-79)
I am saying no to the how.

Let me rephrase - for Airflow 2 users who runs fab provider in any future version (with this PR included) will there be migration scripts running as part of the provider? If the answer is yes then I have a concern here.

**Comment by ephraimbuddy** [comment]
> > These changes are part of AIP-79, do you think we still need another mailing thread on it?
> 
> To clarify, I am not saying no to the what (the merit of AIP-79) I am saying no to the how.
> 
> Let me rephrase - for Airflow 2 users who runs fab provider in any future version (with this PR included) will there be migration scripts running as part of the provider? If the answer is yes then I have a concern here.

There is no migration script currently. However, if anyone has migration to add, they can add it on the FAB provider side. What this PR intends to achieve is for users to add their migrations on the FAB side and have a way for airflow to run such migrations. Even before this PR, a pre-commit forbids adding FAB migration to the core.

**Comment by eladkal** [comment]
>There is no migration script currently. However, if anyone has migration to add, they can add it on the FAB provider side. What this PR intends to achieve is for users to add their migrations on the FAB side and have a way for airflow to run such migrations. 

Great. Then there are no immediate concerns from my side.
I do have several concerns regarding the first migration that we will add and I will raise it when it's relevant

**Review by eladkal — APPROVED** [review]

**Comment by ephraimbuddy** [comment]
I would appreciate some help on this CI issue cc @potiuk @jedcunningham

**Comment by potiuk** [comment]
> I would appreciate some help on this CI issue cc @potiuk @jedcunningham

Which one :) ?

**Comment by ephraimbuddy** [comment]
> > I would appreciate some help on this CI issue cc @potiuk @jedcunningham
> 
> Which one :) ?

Honestly, it's a lot. Something is fundamentally wrong, but I can't find it. It looks related to logging

File: airflow/providers/fab/migrations/env.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+# this is the Alembic Config object, which provides
+# access to the values within the .ini file in use.
+config = context.config
+
+version_table = FABDBManager.version_table_name
+
+# Interpret the config file for Python logging.
+# This line sets up loggers basically.
+if config.config_file_name is not None:
+    fileConfig(config.config_file_name)
```
Comment (jedcunningham) [inline airflow/providers/fab/migrations/env.py]: @ephraimbuddy this is the line that causes the caplog failures! Not exactly sure why, but if I comment this out they go away.

File: airflow/config_templates/config.yml [hunk 3]
```diff
@@ -720,6 +720,14 @@ database:
...
       example: ~
       default: "True"
+    external_db_managers:
+      description: |
+        List of DB managers to use to migrate external tables in airflow database. The managers must inherit
+        from BaseDBManager
+      version_added: 3.0.0
+      type: string
+      example: ~
+      default: "airflow.providers.fab.auth_manager.models.db.FABDBManager"
```
Comment (jedcunningham) [inline airflow/config_templates/config.yml:730]: The default should be null I think. We don't want things to blow up if you don't have the FAB provider installed.

File: airflow/utils/db_manager.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    def __new__(cls, name, bases, dct):
+        """Check that subclasses are setting the required attributes."""
+        required_attrs = ["metadata", "migration_dir", "alembic_file", "version_table_name"]
+        for attr in required_attrs:
+            if attr not in dct:
+                raise AttributeError(f"{name} is missing required attribute: {attr}")
+        return super().__new__(cls, name, bases, dct)
+
+
+class BaseDBManager(LoggingMixin, metaclass=AttributeCheckerMeta):
```
Comment (jedcunningham) [inline airflow/utils/db_manager.py]: Can this be an AbstractBaseClass instead of rolling our own AttributeCheckMeta?

File: airflow/utils/db_manager.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+    def upgradedb(self, to_version=None, from_version=None, show_sql_only=False):
+        """Upgrade the database."""
+        from alembic import command
+
+        config = self.get_alembic_config()
+        command.upgrade(config, revision=to_version or "heads", sql=show_sql_only)
+
+    def downgradedb(self, to_version, from_version=None, show_sql_only=False):
+        """Downgrade the database."""
```
Comment (jedcunningham) [inline airflow/utils/db_manager.py:91]: Was the downgrade skipped intentionally?

File: airflow/providers/fab/migrations/env.py (hunk 2 above)
Comment (ephraimbuddy) [reply airflow/providers/fab/migrations/env.py]: Wow. Pushing up a change soon!

File: airflow/config_templates/config.yml (hunk 3 above)
Comment (ephraimbuddy) [reply airflow/config_templates/config.yml:730]: Should we do this later when we have a replacement for FAB since auth is still tied to FAB?

File: airflow/utils/db_manager.py (hunk 4 above)
Comment (ephraimbuddy) [reply airflow/utils/db_manager.py]: Done

File: airflow/utils/db_manager.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/utils/db_manager.py:91]: Yes. I think it's better implemented by the subclasses

File: airflow/config_templates/config.yml (hunk 3 above)
Comment (jedcunningham) [reply airflow/config_templates/config.yml:730]: Hmm, yeah, good call.

File: airflow/utils/db_manager.py (hunk 5 above)
Comment (jedcunningham) [reply airflow/utils/db_manager.py:91]: Cool, should probably be an abstract method then.

File: .github/workflows/basic-tests.yml [hunk 6]
```diff
@@ -154,8 +154,11 @@ jobs:
         env:
           HATCH_ENV: "test"
         working-directory: ./clients/python
+      - name: "Prepare FAB provider packages: wheel"
```
Comment (potiuk) [inline .github/workflows/basic-tests.yml:157]: Nice.

**Comment by potiuk** [comment]
A lot to look at ... tomorrow.

**Comment by uranusjr** [comment]
```
WARNING: Requirement 'dist/apache_airflow_providers_fab-*.whl' looks like a filename, but the file does not exist
ERROR: apache_airflow_providers_fab-*.whl is not a valid wheel filename.
```

Looks like there’s a problem creating the new provider package?

**Comment by ephraimbuddy** [comment]
> ```
> WARNING: Requirement 'dist/apache_airflow_providers_fab-*.whl' looks like a filename, but the file does not exist
> ERROR: apache_airflow_providers_fab-*.whl is not a valid wheel filename.
> ```
> 
> Looks like there’s a problem creating the new provider package?

A tag with the current fab version was just created, that makes it fail.:

```
The 'final' tag providers-fab/1.3.0 exists. Skipping the package.


Summary of prepared packages:


No packages prepared!
```

Need to find a way to prepare it

**Comment by potiuk** [comment]
I do not know that much of alembic but this one looks good to me. I think we will yet heave to turn FAB provider into something more (airflow UI plugin)

**Review by potiuk — APPROVED** [review]
